### DISCUSSION TITLE
Suggestion: Make better use of the knowledge you have of intellisense

### DISCUSSION THREAD

**Discussion by drizztdourden08** [community/community#11460] [Copilot Conversations]
Sometimes, copilot will suggestion completing something with totally different name from what already exist but very similar. I'm not sure if it currently able to look through the intelisense schemas but it would probably be useful and very helpful to do so.

**Comment by KTibow** [comment]
Well, Copilot is a completion AI. You feed it some code, and it predicts what happens next.

You can see what data it's given in Terminal > Output > GitHub Copilot. How exactly do you think you could feed it the IntelliSense data, as code?

**Comment by drizztdourden08** [comment]
Of course it learn from the code itself. But adding the ability to check the current context outside of the written code and using intellisense data/prediction to add up to the learning It does.

even though it does a pretty good job at generating code, it often doesn’t use the schematics written in the solution itself to generate that same code.

let’s say I have a function:
function GetMyBigPotatos(int potatoX, int potatoY) { do something }

And that it predict the use of that same function somewhere else in the code. Well, it might give you « GetAllMyBigPotatos(2,4, green); » for example or something else. It might add parameters that exist nowhere in the code. Sometimes, it’ll get it right.

Then, that’s just a matter of going in the code and modifying those bits.

And it’ll do that to all kind of objects and structures, including for example if the function is within a class « Potato » for example.

My point here is that intelisense is able to easily detect the function name as a suggestion when I do it manually so I don’t make the mistake myself. Adding that same ability to copilot so it can detect variables name, function names or structures and gove us better snippet of code would be way better and reduce the number of times I have to correct stupid typo or that a snippet is literally unusable because it decided to add a bunch of parameters that do mot exist in the original function.

**Reply by KTibow** [reply]
Could you give an example of the code you're feeding Copilot? *For me* Copilot seems to be able to accurately use previously declared functions. Remember, you can't call an AI bad and hope that it'll improve out of nowhere.

**Reply by drizztdourden08** [reply]
I have never said it was bad. In fact, I love it very much. I only think it can be improved upon still.

From what I can observe, it might be due to the amount of code. This project I have been using it in has around 4000 lines so far and is divided in a lot of subfiles.

I remember it being less of an issue at the beginning of the project.

I am at a debugging stage but as soon as I get an example, I’ll post it here. Hopefully tonight.

**Reply by KTibow** [reply]
> This project I have been using it in has around 4000 lines

That sounds like the problem. I'm pretty sure Copilot has a cap on the amount of context sent.

This sounds possible though, given that sometimes Copilot generates a `# Compare this snippet from otherfile ... code context`, it definitely could be given more info.

**Comment by drizztdourden08** [comment]
Here's an example I've come across just now:

![image](https://user-images.githubusercontent.com/36481532/154394910-5bb51ab1-6f78-490b-bf06-2aee9e2f4909.png)

There's a few things that could help if it was checking like I suggested:

- "Value" doesn't even exist in this class (object)
- (Left) It is definitly not of the type "ZoningType". Neither is any of the parent objects.
- The Enumaration "ZoningType" does not contain an element named "Floor"

There isn't even any of them containing the word floor:

    internal enum ZoningType
    {
        Empty,
        Boundarywall,
        MainZone,
        MainZoneWall,
        MainZoneDoor,
        SubZone,
        SubZoneWall,
        SubZoneDoor,
        SubZoneCorridor,
        StartingPoint = 98,
        Mixed = 99
    }

To be fair, the comment before was asking for it and two of them actually reflect floor tiles (MainZone and SubZone) so I understand where it's coming from but checking the actual structure of the suggested object might have led to a better snippet here.

Hope that makes it clearer.

For the code itself. The project contains 2064 lines split across 11 files.

**Reply by KTibow** [reply]
I see. The hard thing is that you can't pack all type definitions for a whole project into some code for Codex to complete. From [OpenAI's site](https://openai.com/blog/openai-codex/):
> It has a memory of 14KB for Python code, compared to GPT-3 which has only 4KB—so it can take into account over 3x as much contextual information while performing any task.

If you had 4000 lines of code and each was on average 60 characters long, that would be 240KB of context, which would be way out of scope for Codex.

**Reply by drizztdourden08** [reply]
I see. Not sure how the backend works and why there's a limit but I'm guessing this is set in stone for now.

Wouldn't that be easier to store part of the data from intelisense to learn from instead of the whole code?

I know intelisense might not reach all languages and IDE out there and that it might be out of scope but at some point, it might be easier to analyze the code and keep the header of functions, classes and so on only to make prediction instead of the whole thing if it's too much on memory? 

I'm just thinking maybe there's a way to combine both so big project aren't stopped from using it.

Maybe an option to limit the suggestion to this particular file? (only as an option though. It doesn't solve the issue but might help in some situation)

At this point, I'm at 2000 and not 4000. I don't think it's out of the reach of the common project so this might not be an issue only for me but for many in the future, isn't it?

**Reply by KTibow** [reply]
Well from this point I'm not sure how you could feed Codex useful data. Probably someone who spends their job working on Copilot would know more about this. It definitely sounds helpful, though.

**Comment by Rizhiy** [comment]
To achieve this copilot probably has to be retrained such that it is given better information about the current instance at the time of prediction. It would have to be something like:
1. Get context from intellisense
2. Put relevant signatures into the prompt
3. Get predictions.

Currently, it seems to basically be working in simple text-prediction mode, which fails frequently on large projects.
